File: lib/video/schema.ts
import { z } from "zod";

import { collectVideoValidationIssues } from "@/lib/video/validation";

const hexColorPattern = /^#(?:[0-9a-f]{3}|[0-9a-f]{6})$/i;
const fin = z.number().finite();
const pos = z.number().positive();
const nn = z.number().nonnegative();

export const videoHexColorSchema = z
  .string()
  .trim()
  .regex(hexColorPattern, "Colors must use hex notation.");

export const videoAnchorSchema = z.enum([
  "top-left",
  "top-center",
  "top-right",
  "center-left",
  "center",
  "center-right",
  "bottom-left",
  "bottom-center",
  "bottom-right",
]);

export const videoEasingSchema = z.enum([
  "linear",
  "ease-in",
  "ease-out",
  "ease-in-out",
  "ease-in-expo",
  "ease-out-expo",
  "ease-in-back",
  "ease-out-back",
  "spring",
]);

export const videoIconLineCapSchema = z.enum(["butt", "round", "square"]);
export const videoIconLineJoinSchema = z.enum(["bevel", "miter", "round"]);
export const videoArrowPositionSchema = z.enum([
  "above",
  "below",
  "left",
  "right",
]);
export const videoTextAlignSchema = z.enum(["center", "left", "right"]);
export const videoStackDirectionSchema = z.enum(["horizontal", "vertical"]);
export const videoStackAlignSchema = z.enum(["start", "center", "end"]);

// ---------------------------------------------------------------------------
// Base node — shared layout + transform properties (no id; dict key = id)
// ---------------------------------------------------------------------------

const videoNodeBaseSchema = z.object({
  opacity: fin.optional(),
  rotate: fin.optional(),
  scale: fin.optional(),
  scaleX: fin.optional(),
  scaleY: fin.optional(),
  skewX: fin.optional(),
  skewY: fin.optional(),
  x: fin.optional(),
  y: fin.optional(),
  zIndex: z.number().int().optional(),
});

// ---------------------------------------------------------------------------
// Leaf node types (no wrapper / container nodes)
// ---------------------------------------------------------------------------

export const videoRectNodeSchema = videoNodeBaseSchema
  .extend({
    cornerRadius: nn.optional(),
    fill: videoHexColorSchema.optional(),
    height: pos,
    stroke: videoHexColorSchema.optional(),
    strokeWidth: nn.optional(),
    type: z.literal("rect"),
    width: pos,
  })
  .strict();

export const videoPointSchema = z
  .object({
    x: fin,
    y: fin,
  })
  .strict();

export const videoArrowEndpointRefSchema = z
  .object({
    anchor: videoAnchorSchema.optional(),
    node: z.string().trim().min(1),
  })
  .strict();

export const videoArrowEndpointSchema = z.union([
  videoPointSchema,
  videoArrowEndpointRefSchema,
]);

export const videoArrowNodeSchema = videoNodeBaseSchema
  .extend({
    from: videoArrowEndpointSchema.optional(),
    gap: nn.optional(),
    headSize: pos.optional(),
    length: pos.optional(),
    position: videoArrowPositionSchema.optional(),
    stroke: videoHexColorSchema.optional(),
    strokeWidth: nn.optional(),
    target: z.string().trim().min(1).optional(),
    to: videoArrowEndpointSchema.optional(),
    type: z.literal("arrow"),
  })
  .strict();

export const videoTextNodeSchema = videoNodeBaseSchema
  .extend({
    color: videoHexColorSchema.optional(),
    fontFamily: z.string().trim().min(1).optional(),
    fontWeight: z.union([fin, z.string().trim().min(1)]).optional(),
    lineHeight: pos.optional(),
    maxWidth: pos.optional(),
    size: pos.optional(),
    text: z.string().trim().min(1),
    textAlign: videoTextAlignSchema.optional(),
    type: z.literal("text"),
  })
  .strict();

const videoIconPointSchema = z.tuple([fin, fin]);

export const videoIconPathPrimitiveSchema = z
  .object({
    d: z.string().trim().min(1),
    type: z.literal("path"),
  })
  .strict();

export const videoIconCirclePrimitiveSchema = z
  .object({
    cx: fin,
    cy: fin,
    r: pos,
    type: z.literal("circle"),
  })
  .strict();

export const videoIconLinePrimitiveSchema = z
  .object({
    type: z.literal("line"),
    x1: fin,
    x2: fin,
    y1: fin,
    y2: fin,
  })
  .strict();

export const videoIconPolylinePrimitiveSchema = z
  .object({
    points: z.array(videoIconPointSchema).min(2),
    type: z.literal("polyline"),
  })
  .strict();

export const videoIconPolygonPrimitiveSchema = z
  .object({
    points: z.array(videoIconPointSchema).min(3),
    type: z.literal("polygon"),
  })
  .strict();

export const videoIconRectPrimitiveSchema = z
  .object({
    height: pos,
    rx: nn.optional(),
    ry: nn.optional(),
    type: z.literal("rect"),
    width: pos,
    x: fin.optional(),
    y: fin.optional(),
  })
  .strict();

export const videoIconPrimitiveSchema = z.discriminatedUnion("type", [
  videoIconPathPrimitiveSchema,
  videoIconCirclePrimitiveSchema,
  videoIconLinePrimitiveSchema,
  videoIconPolylinePrimitiveSchema,
  videoIconPolygonPrimitiveSchema,
  videoIconRectPrimitiveSchema,
]);

// Full engine-facing schema: requires resolved SVG elements.
export const videoIconNodeSchema = videoNodeBaseSchema
  .extend({
    absoluteStrokeWidth: z.boolean().optional(),
    elements: z.array(videoIconPrimitiveSchema).min(1),
    fill: videoHexColorSchema.optional(),
    height: pos,
    lineCap: videoIconLineCapSchema.optional(),
    lineJoin: videoIconLineJoinSchema.optional(),
    stroke: videoHexColorSchema.optional(),
    strokeWidth: nn.optional(),
    type: z.literal("icon"),
    viewportHeight: pos.optional(),
    viewportWidth: pos.optional(),
    width: pos,
  })
  .strict();

// AI-output schema: references icons by Lucide name; the server resolves
// names to elements before passing the description to the Rust engine.
export const videoAiIconNodeSchema = videoNodeBaseSchema
  .extend({
    fill: videoHexColorSchema.optional(),
    height: pos,
    lineCap: videoIconLineCapSchema.optional(),
    lineJoin: videoIconLineJoinSchema.optional(),
    name: z.string().trim().min(1),
    stroke: videoHexColorSchema.optional(),
    strokeWidth: nn.optional(),
    type: z.literal("icon"),
    width: pos,
  })
  .strict();

export const videoCenterNodeSchema = videoNodeBaseSchema
  .extend({
    children: z.array(z.string().min(1)).length(1),
    height: pos.optional(),
    type: z.literal("center"),
    width: pos.optional(),
  })
  .strict();

export const videoAlignNodeSchema = videoNodeBaseSchema
  .extend({
    children: z.array(z.string().min(1)).length(1),
    height: pos.optional(),
    padding: nn.optional(),
    position: videoAnchorSchema,
    type: z.literal("align"),
    width: pos.optional(),
  })
  .strict();

export const videoStackNodeSchema = videoNodeBaseSchema
  .extend({
    align: videoStackAlignSchema.optional(),
    children: z.array(z.string().min(1)).min(1),
    direction: videoStackDirectionSchema,
    gap: nn.optional(),
    height: pos.optional(),
    type: z.literal("stack"),
    width: pos.optional(),
  })
  .strict();

const videoRepeatTemplateSchema = z.union([
  videoArrowNodeSchema,
  videoAiIconNodeSchema,
  videoRectNodeSchema,
  videoTextNodeSchema,
]);

export const videoRepeatNodeSchema = z
  .object({
    colStep: videoPointSchema.optional(),
    cols: z.number().int().positive().max(32),
    origin: videoPointSchema.optional(),
    rowStep: videoPointSchema.optional(),
    rows: z.number().int().positive().max(32),
    template: videoRepeatTemplateSchema,
    type: z.literal("repeat"),
  })
  .strict();

// ---------------------------------------------------------------------------
// Node unions
// ---------------------------------------------------------------------------

// Engine node union: all types with fully-resolved payloads.
export const videoNodeSchema = z.discriminatedUnion("type", [
  videoAlignNodeSchema,
  videoArrowNodeSchema,
  videoCenterNodeSchema,
  videoIconNodeSchema,
  videoRectNodeSchema,
  videoStackNodeSchema,
  videoTextNodeSchema,
]);

// AI output node union: identical to the engine union except icon nodes carry
// a Lucide name instead of raw elements. The server resolves names before
// serialising to Rust.
export const videoAiNodeSchema = z.discriminatedUnion("type", [
  videoAlignNodeSchema,
  videoArrowNodeSchema,
  videoCenterNodeSchema,
  videoAiIconNodeSchema,
  videoRectNodeSchema,
  videoRepeatNodeSchema,
  videoStackNodeSchema,
  videoTextNodeSchema,
]);

// ---------------------------------------------------------------------------
// Timeline event — centralized animation per scene
// ---------------------------------------------------------------------------

export const videoTimelineEventSchema = z
  .object({
    action: z.enum(["draw"]).optional(),
    at: nn,
    color: videoHexColorSchema.optional(),
    cornerRadius: nn.optional(),
    drawProgress: z.number().min(0).max(1).optional(),
    dur: pos.optional(),
    dx: fin.optional(),
    dy: fin.optional(),
    ease: videoEasingSchema.optional(),
    fill: videoHexColorSchema.optional(),
    height: pos.optional(),
    opacity: fin.optional(),
    rotate: fin.optional(),
    scale: fin.optional(),
    scaleX: fin.optional(),
    scaleY: fin.optional(),
    size: pos.optional(),
    skewX: fin.optional(),
    skewY: fin.optional(),
    stroke: videoHexColorSchema.optional(),
    strokeWidth: nn.optional(),
    target: z.union([z.string().min(1), z.array(z.string().min(1)).min(1)]),
    width: pos.optional(),
    x: fin.optional(),
    y: fin.optional(),
  })
  .strict();

// ---------------------------------------------------------------------------
// Scenes
// ---------------------------------------------------------------------------

export const videoAiSceneSchema = z
  .object({
    background: videoHexColorSchema.optional(),
    duration: pos,
    id: z.string().trim().min(1),
    nodes: z.record(z.string(), videoAiNodeSchema),
    timeline: z.array(videoTimelineEventSchema).optional(),
  })
  .strict();

export const videoSceneSchema = z
  .object({
    background: videoHexColorSchema.optional(),
    duration: z.number().int().positive(),
    id: z.string().trim().min(1),
    nodes: z.record(z.string(), videoNodeSchema),
    startFrame: z.number().int().nonnegative(),
    timeline: z.array(videoTimelineEventSchema).optional(),
  })
  .strict();

// ---------------------------------------------------------------------------
// Top-level descriptors
// ---------------------------------------------------------------------------

export const videoDescriptionSchema = z
  .object({
    background: videoHexColorSchema.optional(),
    fps: pos,
    height: z.number().int().positive(),
    scenes: z.array(videoSceneSchema).min(1),
    width: z.number().int().positive(),
  })
  .strict()
  .superRefine((desc, ctx) => {
    for (const issue of collectVideoValidationIssues(desc)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: issue.message,
        path: issue.path,
      });
    }
  });

export const videoAiOutputSchema = z
  .object({
    background: videoHexColorSchema.optional(),
    scenes: z.array(videoAiSceneSchema).min(1),
  })
  .strict();

File: engine/src/schema.rs
#![allow(dead_code)]

use indexmap::IndexMap;
use serde::Deserialize;

#[derive(Debug, Clone, Copy, Deserialize)]
pub enum Anchor {
    #[serde(rename = "top-left")]
    TopLeft,
    #[serde(rename = "top-center")]
    TopCenter,
    #[serde(rename = "top-right")]
    TopRight,
    #[serde(rename = "center-left")]
    CenterLeft,
    #[serde(rename = "center")]
    Center,
    #[serde(rename = "center-right")]
    CenterRight,
    #[serde(rename = "bottom-left")]
    BottomLeft,
    #[serde(rename = "bottom-center")]
    BottomCenter,
    #[serde(rename = "bottom-right")]
    BottomRight,
}

#[derive(Debug, Clone, Copy, Deserialize)]
pub enum ArrowPosition {
    #[serde(rename = "above")]
    Above,
    #[serde(rename = "below")]
    Below,
    #[serde(rename = "left")]
    Left,
    #[serde(rename = "right")]
    Right,
}

#[derive(Debug, Clone, Copy, Deserialize)]
pub enum Easing {
    #[serde(rename = "linear")]
    Linear,
    #[serde(rename = "ease-in")]
    EaseIn,
    #[serde(rename = "ease-out")]
    EaseOut,
    #[serde(rename = "ease-in-out")]
    EaseInOut,
    #[serde(rename = "ease-in-expo")]
    EaseInExpo,
    #[serde(rename = "ease-out-expo")]
    EaseOutExpo,
    #[serde(rename = "ease-in-back")]
    EaseInBack,
    #[serde(rename = "ease-out-back")]
    EaseOutBack,
    #[serde(rename = "spring")]
    Spring,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum StackAlign {
    Start,
    Center,
    End,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum StackDirection {
    Horizontal,
    Vertical,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TextAlign {
    Left,
    Center,
    Right,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum IconLineCap {
    Butt,
    Round,
    Square,
}

#[derive(Debug, Clone, Copy, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum IconLineJoin {
    Bevel,
    Miter,
    Round,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum FontWeight {
    Numeric(f64),
    Named(String),
}

#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum EventTarget {
    Single(String),
    Multiple(Vec<String>),
}

impl EventTarget {
    pub fn contains(&self, id: &str) -> bool {
        match self {
            Self::Single(s) => s == id,
            Self::Multiple(v) => v.iter().any(|s| s == id),
        }
    }
}

// ---------------------------------------------------------------------------
// Top-level
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct VideoDescription {
    pub fps: f64,
    pub width: u32,
    pub height: u32,
    #[serde(default)]
    pub background: Option<String>,
    pub scenes: Vec<SceneEntry>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SceneEntry {
    pub id: String,
    #[serde(default)]
    pub background: Option<String>,
    pub duration: u32,
    pub start_frame: u32,
    pub nodes: IndexMap<String, Node>,
    #[serde(default)]
    pub timeline: Vec<TimelineEvent>,
}

// ---------------------------------------------------------------------------
// Shared node layout/transform fields
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NodeBase {
    pub opacity: Option<f64>,
    pub rotate: Option<f64>,
    pub scale: Option<f64>,
    pub scale_x: Option<f64>,
    pub scale_y: Option<f64>,
    pub skew_x: Option<f64>,
    pub skew_y: Option<f64>,
    pub x: Option<f64>,
    pub y: Option<f64>,
    pub z_index: Option<i32>,
}

// ---------------------------------------------------------------------------
// Node variants — tagged union on "type"
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type")]
pub enum Node {
    #[serde(rename = "align")]
    Align(AlignNode),
    #[serde(rename = "arrow")]
    Arrow(ArrowNode),
    #[serde(rename = "center")]
    Center(CenterNode),
    #[serde(rename = "functionGraph")]
    FunctionGraph(FunctionGraphNode),
    #[serde(rename = "icon")]
    Icon(IconNode),
    #[serde(rename = "parametricGraph")]
    ParametricGraph(ParametricGraphNode),
    #[serde(rename = "rect")]
    Rect(RectNode),
    #[serde(rename = "stack")]
    Stack(StackNode),
    #[serde(rename = "text")]
    Text(TextNode),
}

impl Node {
    pub fn base(&self) -> &NodeBase {
        match self {
            Self::Align(n) => &n.base,
            Self::Arrow(n) => &n.base,
            Self::Center(n) => &n.base,
            Self::FunctionGraph(n) => &n.base,
            Self::Icon(n) => &n.base,
            Self::ParametricGraph(n) => &n.base,
            Self::Rect(n) => &n.base,
            Self::Stack(n) => &n.base,
            Self::Text(n) => &n.base,
        }
    }

    pub fn children(&self) -> &[String] {
        match self {
            Self::Align(n) => &n.children,
            Self::Center(n) => &n.children,
            Self::Stack(n) => &n.children,
            _ => &[],
        }
    }

    pub fn is_layout(&self) -> bool {
        matches!(self, Self::Align(_) | Self::Center(_) | Self::Stack(_))
    }
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArrowNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub from: Option<ArrowEndpoint>,
    pub to: Option<ArrowEndpoint>,
    pub target: Option<String>,
    pub position: Option<ArrowPosition>,
    pub gap: Option<f64>,
    pub length: Option<f64>,
    pub stroke: Option<String>,
    pub stroke_width: Option<f64>,
    pub head_size: Option<f64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(untagged)]
pub enum ArrowEndpoint {
    Point(ArrowPoint),
    NodeRef(ArrowEndpointTarget),
}

#[derive(Debug, Clone, Deserialize)]
pub struct ArrowPoint {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Copy, Deserialize)]
pub struct GraphPoint {
    pub x: f64,
    pub y: f64,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArrowEndpointTarget {
    pub node: String,
    pub anchor: Option<Anchor>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CenterNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub children: Vec<String>,
    pub width: Option<f64>,
    pub height: Option<f64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AlignNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub children: Vec<String>,
    pub position: Anchor,
    pub padding: Option<f64>,
    pub width: Option<f64>,
    pub height: Option<f64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StackNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub children: Vec<String>,
    pub direction: StackDirection,
    pub gap: Option<f64>,
    pub align: Option<StackAlign>,
    pub width: Option<f64>,
    pub height: Option<f64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RectNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub width: f64,
    pub height: f64,
    pub fill: Option<String>,
    pub stroke: Option<String>,
    pub stroke_width: Option<f64>,
    pub corner_radius: Option<f64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FunctionGraphNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub width: f64,
    pub height: f64,
    pub points: Vec<GraphPoint>,
    pub color: Option<String>,
    pub stroke_width: Option<f64>,
    pub show_axes: Option<bool>,
    pub show_grid: Option<bool>,
    pub draw_progress: Option<f64>,
    pub x_range: Option<[f64; 2]>,
    pub y_range: Option<[f64; 2]>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParametricGraphNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub width: f64,
    pub height: f64,
    pub points: Vec<GraphPoint>,
    pub color: Option<String>,
    pub stroke_width: Option<f64>,
    pub draw_progress: Option<f64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TextNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub text: String,
    pub color: Option<String>,
    pub font_family: Option<String>,
    pub font_weight: Option<FontWeight>,
    pub line_height: Option<f64>,
    pub max_width: Option<f64>,
    pub size: Option<f64>,
    pub text_align: Option<TextAlign>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IconNode {
    #[serde(flatten)]
    pub base: NodeBase,
    pub width: f64,
    pub height: f64,
    pub viewport_width: Option<f64>,
    pub viewport_height: Option<f64>,
    pub stroke: Option<String>,
    pub fill: Option<String>,
    pub stroke_width: Option<f64>,
    pub absolute_stroke_width: Option<bool>,
    pub line_cap: Option<IconLineCap>,
    pub line_join: Option<IconLineJoin>,
    #[serde(default)]
    pub elements: Vec<IconPrimitive>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type")]
pub enum IconPrimitive {
    #[serde(rename = "path")]
    Path(IconPathPrimitive),
    #[serde(rename = "circle")]
    Circle(IconCirclePrimitive),
    #[serde(rename = "line")]
    Line(IconLinePrimitive),
    #[serde(rename = "polyline")]
    Polyline(IconPolylinePrimitive),
    #[serde(rename = "polygon")]
    Polygon(IconPolygonPrimitive),
    #[serde(rename = "rect")]
    Rect(IconRectPrimitive),
}

#[derive(Debug, Clone, Deserialize)]
pub struct IconPathPrimitive {
    pub d: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IconCirclePrimitive {
    pub cx: f64,
    pub cy: f64,
    pub r: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IconLinePrimitive {
    pub x1: f64,
    pub y1: f64,
    pub x2: f64,
    pub y2: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IconPolylinePrimitive {
    pub points: Vec<(f64, f64)>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IconPolygonPrimitive {
    pub points: Vec<(f64, f64)>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IconRectPrimitive {
    pub x: Option<f64>,
    pub y: Option<f64>,
    pub width: f64,
    pub height: f64,
    pub rx: Option<f64>,
    pub ry: Option<f64>,
}

// ---------------------------------------------------------------------------
// Timeline events
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TimelineEvent {
    pub target: EventTarget,
    pub at: f64,
    pub dur: Option<f64>,
    pub ease: Option<Easing>,
    pub action: Option<String>,
    // Numeric animatable properties
    pub opacity: Option<f64>,
    pub x: Option<f64>,
    pub y: Option<f64>,
    pub dx: Option<f64>,
    pub dy: Option<f64>,
    pub width: Option<f64>,
    pub height: Option<f64>,
    pub rotate: Option<f64>,
    pub scale: Option<f64>,
    pub scale_x: Option<f64>,
    pub scale_y: Option<f64>,
    pub skew_x: Option<f64>,
    pub skew_y: Option<f64>,
    pub corner_radius: Option<f64>,
    pub stroke_width: Option<f64>,
    pub size: Option<f64>,
    pub draw_progress: Option<f64>,
    // Color animatable properties
    pub fill: Option<String>,
    pub stroke: Option<String>,
    pub color: Option<String>,
}

impl TimelineEvent {
    pub fn get_num(&self, prop: &str) -> Option<f64> {
        match prop {
            "opacity" => self.opacity,
            "x" => self.x,
            "y" => self.y,
            "dx" => self.dx,
            "dy" => self.dy,
            "width" => self.width,
            "height" => self.height,
            "rotate" => self.rotate,
            "scale" => self.scale,
            "scaleX" => self.scale_x,
            "scaleY" => self.scale_y,
            "skewX" => self.skew_x,
            "skewY" => self.skew_y,
            "cornerRadius" => self.corner_radius,
            "strokeWidth" => self.stroke_width,
            "size" => self.size,
            "drawProgress" => self.draw_progress,
            _ => None,
        }
    }

    pub fn get_color(&self, prop: &str) -> Option<&str> {
        match prop {
            "fill" => self.fill.as_deref(),
            "stroke" => self.stroke.as_deref(),
            "color" => self.color.as_deref(),
            _ => None,
        }
    }
}
